src/PublicApiAnalyzers/Core/Analyzers/DeclarePublicApiAnalyzer.Impl.cs

```diff
@@ -875,17 +879,37 @@ public override bool VisitPointerType(IPointerTypeSymbol symbol)
                     return Visit(symbol.PointedAtType);
                 }
 
+                /// <summary>This only checks the use of a type parameter. We're checking their definition (looking at type constraints) elsewhere.</summary>
                 public override bool VisitTypeParameter(ITypeParameterSymbol symbol)
                 {
-                    if (symbol.HasReferenceTypeConstraint() && symbol.ReferenceTypeConstraintNullableAnnotation() == NullableAnnotation.None)
+                    if (symbol.IsReferenceType &&
+                        symbol.NullableAnnotation() == NullableAnnotation.None)
+                    {
+                        // Example:
+                        // I<TReferenceType~>
+                        return true;
+                    }
+
+                    return false;
+                }
+
+                /// <summary>This is checking the definition of a type parameter (as opposed to its usage).</summary>
+                public static bool CheckTypeParameterConstraints(ITypeParameterSymbol symbol)
+                {
+                    if (symbol.HasReferenceTypeConstraint() &&
+                        symbol.ReferenceTypeConstraintNullableAnnotation() == NullableAnnotation.None)
                     {
+                        // where T : class~
                         return true;
                     }
 
                     foreach (var constraintType in symbol.ConstraintTypes)
```

Comment:
Could also be refactored as a `.Any()` call.

Reply:
I prefer the explicit loop style. One reason is it's easier to step through.